### PR TITLE
[17549] Initial Work Package Creation in Details

### PR DESCRIPTION
**Pull request by floriank**
The original PR was #3124 by @0xF013.

This PR is reopenend against `release/4.3`. The checklist from the original:
- [x] fix autocomplete noscope
- [x] refactor or remove task dropdown menu controller
- [x] fix refresh for /create url
- [x] fix annoying focus on each element on create mode
- [x] focus to title on create toggle
- [x] new work package header in german
    button_save_all: "Save"
    button_save_and_send_all: "Save and send email"
    button_cancel_all: "Cancel"
- [x] column arrows position bug
- [x] time group is now always empty
- [x] payload API title or name
- [x] type switch during creation

---

Work Package: https://community.openproject.org/work_packages/17549

The change introduces the ability to create new work packages from the details pane. The change is a draft preparing to fully implement the workflow in the wake of `4.3`

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://floriank-feature-17549-initial-wp-create-frontend.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://floriank-feature-17549-initial-wp-create-frontend.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by floriank** [comment]
Ready for review imho.

**Comment by floriank** [comment]
I have no idea why the coverage drops here. Travis seems to like my test fixing/mock introducing

File: app/controllers/work_packages/auto_completes_controller.rb [hunk 1]
```diff
@@ -30,43 +30,65 @@
...
+    nil
+  end
+
+  def determine_scope
+    project = find_project
+
+    if params[:scope] == 'relatable'
+      return nil unless project
+
+      Setting.cross_project_work_package_relations? ? WorkPackage.scoped : project.work_packages
```
Comment (myabc) [inline app/controllers/work_packages/auto_completes_controller.rb:87]: @NobodysNightmare `WorkPackages.scoped` is deprecated in Rails 4.

Comment (NobodysNightmare) [inline app/controllers/work_packages/auto_completes_controller.rb:87]: @myabc I know (see my commit message: https://github.com/floriank/openproject/commit/de5bb8f01bfb3c745a269f8c0c3c83b9d958873b).

The problem was that I had no choice (or I didn't know my good options):
- using `:all` seems to be the Rails 4.0+ way, because it ([mostly](http://stackoverflow.com/questions/18198963/with-rails-4-model-scoped-is-deprecated-but-model-all-cant-replace-it)) does what `:scoped` did before
- in Rails 3 `:all` returns a silly `Array`, so I can't use it just now... But `:scoped` does what I want

On another note: I have no idea why this part turns up in the diff of this PR. It should already be included in `release/4.3`

Comment (myabc) [inline app/controllers/work_packages/auto_completes_controller.rb:87]: @NobodysNightmare ohh .. then never mind! I, too, was confused by GitHub showing this in the PR diff . I've actually already dealt with this deprecation in #3194 (see https://github.com/myabc/openproject/commit/0c729dae548a90279c703f820791ef811a43879f#diff-75bc1c1409a39ce317aa768f040e8839L87). Feel free to review :wink:

**Comment by floriank** [comment]
rebased.

**Comment by NobodysNightmare** [comment]
Functional test looks fine. Observations so far:
- after creating a WP it does not appear in the WP list
- Commit buttons are quite far down for high vertical resolutions

Note: Those are only neutral observations, as this PR is intended for an early merge, I don't see them as critical. I will have to find out where to post them nevertheless.

**Comment by floriank** [comment]
> after creating a WP it does not appear in the WP list

This is in my opinion highly desirable, but the easy way to implement this at the moment would probably be to trigger the loading of the current list, **with** it's current filters which might prevent the displaying of the particular WP.

I love the smell of edge cases in the morning.

File: frontend/app/templates/work_packages.list.new.html [hunk 2]
```diff
@@ -0,0 +1,71 @@
+<div cg-busy="vm.loaderPromise" class="work-packages--details-content -create-mode">
+  <div ng-if="vm.workPackage">
+    <h2>{{ ::I18n.t('js.work_packages.create.header') }}</h2>
+
+    <div class="single-attribute subject">
```
Comment (NobodysNightmare) [inline frontend/app/templates/work_packages.list.new.html]: Not quite understanding the purpose of this `div` container. It got my attention, because I saw it in a new CSS rule in `_work_package.sass`.
- Why is it not to be found in the edit view?
- What is a `single-attribute`? Seems like only the subject and the description are "single attributes".
  - for edit mode only the description is a "single attribute"
- Why do we (again!) need to specify that this is the `subject`?
  - the immediate (and only) DOM child already is a `work-packages--details--subject`
  - the DOM grandchild is an `attribute-subject`
- the class `subject` has no prefix and is likely to collide with other CSS rules

File: app/assets/stylesheets/layout/_work_package.sass [hunk 3]
```diff
@@ -105,6 +105,14 @@
   width:      100%
   +allow-vertical-scrolling
   padding:    0 25px 0 25px
+  &.-create-mode
+    top: 0
+    h2
+      margin-top: 0.375em
+    .single-attribute.subject
```
Comment (NobodysNightmare) [inline app/assets/stylesheets/layout/_work_package.sass]: See my comment on the corresponding container below. I'd say one could replace this with the follwoing selector:

``` CSS
.work-package-field.work-packages--details--subject
```

File: frontend/app/templates/work_packages.list.new.html (hunk 2 above)
Comment (floriank) [inline frontend/app/templates/work_packages.list.new.html]: I'll look into it - to be fair, I have not looked out for code quality in @0xF013's code as this should not be the initial (**main**) criteria here. so this is my fault. 

Am I the only on who is bothered by the fact that create and edit view differ that much?

File: frontend/app/api/hal-api-resource.js [hunk 4]
```diff
@@ -62,6 +62,22 @@ module.exports = function HALAPIResource($timeout, $q, PathHelper) {
...
         url: PathHelper.appBasePath + PathHelper.apiV3 + '/' + uri,
       });
+    },
+    setupLink: function(uri, params) {
+      // some monkey-patching here
+      // since the Hyperagent.Link constructor
+      // behaves differently from the links constructed
+      // from the API response
+      HALAPIResource.configure();
+      var url = PathHelper.appBasePath + PathHelper.apiV3 + '/' + uri;
```
Comment (NobodysNightmare) [inline frontend/app/api/hal-api-resource.js:64]: This is looking suspicious:

A hard reference to APIv3 in a neutral looking class. Since it is called "monkey patching" I suppose that this method is called/ used outside of our application too? Because for our application I can only find a single call in the work-package-service.

File: frontend/app/services/work-package-service.js [hunk 5]
```diff
@@ -75,6 +76,47 @@ module.exports = function($http,
...
+        embedded: {},
+        props: {},
+        links: {
+          update: HALAPIResource
+            .setup(PathHelper
+              .projectWorkPackagesFormPath(projectIdentifier)),
+          updateImmediately: HALAPIResource.setupLink(
+            PathHelper.projectWorkPackagesPath(projectIdentifier),
+            { method: 'post' }
+          )
```
Comment (NobodysNightmare) [inline frontend/app/services/work-package-service.js:91]: Those two `HALAPIResource` calls look very suspicious:
- Which use do they have at all? They look like the `createWorkPackage` and `createWorkPackageImmediate` links in a project resource
  - they do **not** look like update links (mind the difference between update and create)
- The `setupLink` helper will prepend an `api/v3` prefix. I am not sure why the correct uri is not passed here in the first place
- From a callers perspective (reading those lines above) I can't tell why `update` needs a `setup` and `updateImmediately` needs a `setupLink`. The names look arbitrary

File: frontend/app/work_packages/controllers/work-package-new-controller.js [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  vm.isEditable = isEditable;
+  vm.hasNiceStar = hasNiceStar;
+  vm.showToggleButton = showToggleButton;
+
+  activate();
+
+  function activate() {
+    EditableFieldsState.forcedEditState = true;
+    vm.loaderPromise = WorkPackageService.initializeWorkPackage($scope.projectIdentifier, {
+      type: PathHelper.apiV3TypePath($stateParams.type)
```
Comment (NobodysNightmare) [inline frontend/app/work_packages/controllers/work-package-new-controller.js]: Disclaimer: I don't know the surrounding/calling code. From a wishful unicorn perspective I would love the following:
- `$stateParams.type` is a type resource as returned by APIv3
- you don't need to use a pathhelper, but simply use `type.links.self.href`

If you say that would require lots of rewriting we should do it later (hope that does not mean "never"). But judging from other code in the frontend, chances are high, that the source of `$stateParams.type` was already a "good"/"full" type and only its id got ripped off. In that case the change would be easy and I'd request it.

File: frontend/app/services/work-package-service.js [hunk 7]
```diff
@@ -25,6 +25,7 @@
 //
 // See doc/COPYRIGHT.rdoc for more details.
 //++
+/* globals URI */
```
Comment (NobodysNightmare) [inline frontend/app/services/work-package-service.js:28]: What?

This line was added during a commit called "address hound complaints".

Either the complaint was

> Please add a useless comment mentioning URIs

or it is some kind of unwanted commit.

File: frontend/app/services/work-package-service.js [hunk 8]
```diff
@@ -75,6 +76,47 @@ module.exports = function($http,
...
   var WorkPackageService = {
+    initializeWorkPackage: function(projectIdentifier, initialData) {
+      var changes = _.clone(initialData);
+      if (initialData.type) {
+        changes._links = changes.links || {};
+        changes._links.type = {
+          href: initialData.type
+        };
+        delete changes.type;
+      }
```
Comment (NobodysNightmare) [inline frontend/app/services/work-package-service.js]: I don't like the mixed reference to `initialData` and `changes` in this if-block. You should probably always refer to `changes`, since this is the copy you just created for use in this method.

File: frontend/app/services/work-package-service.js [hunk 9]
```diff
@@ -75,6 +76,47 @@ module.exports = function($http,
   }
 
   var WorkPackageService = {
+    initializeWorkPackage: function(projectIdentifier, initialData) {
+      var changes = _.clone(initialData);
+      if (initialData.type) {
+        changes._links = changes.links || {};
```
Comment (NobodysNightmare) [inline frontend/app/services/work-package-service.js]: For one I don't get the (in)difference between `_links` and `links`, as well as `_embedded` and`embedded`. I usually try and see which fails.

This gives the line above a gentle, fishy smell. But I can't judge if the smell is okay or not.

File: frontend/app/services/work-package-service.js [hunk 10]
```diff
@@ -75,6 +76,47 @@ module.exports = function($http,
...
+        changes._links.type = {
+          href: initialData.type
+        };
+        delete changes.type;
```
Comment (NobodysNightmare) [inline frontend/app/services/work-package-service.js]: This begs two questions:
- If linked properties like type are misplaced inside `changes` (`type` instead of `_links/type`), what about other linked values (e.g. status and priority)
- Looking at line 82: If we copy `changes.links` into `changes._links` beforehand, why is the type not already part of the changes?

File: frontend/app/services/work-package-service.js [hunk 11]
```diff
@@ -75,6 +76,47 @@ module.exports = function($http,
...
+        }
+      };
+      var options = { ajax: {
+          method: 'POST',
+          headers: {
+            Accept: 'application/hal+json'
+          },
+          data: JSON.stringify(changes),
+          contentType: 'application/json; charset=utf-8'
+        }};
```
Comment (NobodysNightmare) [inline frontend/app/services/work-package-service.js:101]: A note on dryness:

I can't imagine this is the first time we define those options. [They](https://github.com/NobodysNightmare/openproject/blob/2995ad8f53af0c06755d31554f0d3c28e8055baa/frontend/app/services/work-package-service.js#L104-104) [look](https://github.com/NobodysNightmare/openproject/blob/2995ad8f53af0c06755d31554f0d3c28e8055baa/frontend/app/services/work-package-service.js#L227-227) [pretty](https://github.com/NobodysNightmare/openproject/blob/2995ad8f53af0c06755d31554f0d3c28e8055baa/frontend/app/services/work-package-service.js#L255-255) [default](https://github.com/NobodysNightmare/openproject/blob/2995ad8f53af0c06755d31554f0d3c28e8055baa/frontend/app/services/work-package-service.js#L270-270).

File: frontend/app/services/work-package-service.js [hunk 12]
```diff
@@ -221,9 +261,11 @@ module.exports = function($http,
     },
 
     updateWorkPackage: function(workPackage, notify) {
+      console.log(workPackage.links.updateImmediately);
```
Comment (NobodysNightmare) [inline frontend/app/services/work-package-service.js]: Do we usually keep such statements? Looks like printf debugging, but maybe I just don't see the reoccuring value in this log output.

File: frontend/app/services/work-package-service.js [hunk 13]
```diff
@@ -221,9 +261,11 @@ module.exports = function($http,
...
       var options = { ajax: {
-        method: 'PATCH',
-        url: URI(workPackage.links.updateImmediately.href).addSearch('notify', notify).toString(),
+        method: workPackage.links.updateImmediately.props.method,
+        url: URI(workPackage.links.updateImmediately.props.href)
+            .addSearch('notify', notify).toString(),
```
Comment (NobodysNightmare) [inline frontend/app/services/work-package-service.js:260]: I think I see the reason why the WP is initialized with bogus `update` links here.

Hint: The work package form you are working with also always has a link called `commit`. Depending on whether it is a create form or an update form, the links are different. In either case they should also indicate the correct method.
This knowledge might come in handy when refactoring this code.

File: frontend/app/templates/work_packages.list.new.html [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+  <button class="button" ng-click="vm.submit(true)">
+    <i class="icon-yes"></i>
+    <i class="icon-mail"></i>
+    <span class="button--text" ng-bind="::I18n.t('js.inplace.button_save_and_send_all')"></span>
+  </button>
+  <button class="button" ng-click="vm.cancel()">
+    <i class="icon-close"></i>
+    <span class="button--text" ng-bind="::I18n.t('js.inplace.button_cancel_all')"></span>
+  </button>
+</div>
```
Comment (NobodysNightmare) [inline frontend/app/templates/work_packages.list.new.html]: Looks like there is very few reuse between the template for the edit view and the template for the create form. I would have thought that we can share more here...

File: frontend/app/work_packages/controllers/menus/tasks-dropdown-menu-controller.js [hunk 15]
```diff
@@ -27,10 +27,5 @@
 //++
 
 module.exports = function($scope, PathHelper) {
-  $scope.workPackageNewPath = function(typeId) {
-    return PathHelper.staticWorkPackageNewWithParametersPath(
-      $scope.projectIdentifier,
-      { 'type_id': typeId }
-    );
-  };
+
```
Comment (NobodysNightmare) [inline frontend/app/work_packages/controllers/menus/tasks-dropdown-menu-controller.js]: Empty controllers are a thing?

File: frontend/app/work_packages/controllers/work-package-new-controller.js [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+           $timeout,
+           $window,
+           PathHelper,
+           WorkPackagesOverviewService,
+           WorkPackageFieldService,
+           WorkPackageService,
+           EditableFieldsState
+           ) {
+
+  var vm = this,
```
Comment (NobodysNightmare) [inline frontend/app/work_packages/controllers/work-package-new-controller.js]: Just curious: What does `vm` stand for?

File: frontend/app/work_packages/controllers/work-package-new-controller.js [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+        });
+        if (!firstTimeFocused) {
+          firstTimeFocused = true;
+          $timeout(function() {
+            // TODO: figure out a better way to fix the wp table columns bug
+            // where arrows are misplaced when not resizing the window
+            angular.element($window).trigger('resize');
+            angular.element('.work-packages--details--subject .focus-input').focus();
+          });
+        }
```
Comment (NobodysNightmare) [inline frontend/app/work_packages/controllers/work-package-new-controller.js]: If this whole `firstTimeFocused` business is because of a bug in the WP table, why do I not see it in the otherwise identical code for the edit mode? ([identical code here](https://github.com/NobodysNightmare/openproject/blob/7484c1980a57485552215f823364d238f1df9665/frontend/app/work_packages/controllers/details-tab-overview-controller.js#L60-73))

And by extension: Why is this code duplicated?

File: frontend/app/work_packages/controllers/work-package-new-controller.js [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+  }
+
+  function getLabel(field) {
+    return WorkPackageFieldService.getLabel(vm.workPackage, field);
+  }
+
+  function showToggleButton() {
+    return true;
+  }
+};
```
Comment (NobodysNightmare) [inline frontend/app/work_packages/controllers/work-package-new-controller.js:120]: I can understand that there are reasons where you rather copy parts of a method instead of doing the business involved in DRYing it up. (read: Sometimes, under limited circumstances; don't read: best practice)

However, I think there are no reasons to copy over a bunch of helper methods (`isXXX`, `hasXXX`, etc.) from another class that are all completely identical. Here you have several options, among them:
- inheritance from a common base class (if that is a thing in JS)
- using a shared class offering helper methods

File: frontend/app/work_packages/directives/inplace_editor/inplace-editor-edit-pane-directive.js [hunk 19]
```diff
@@ -111,6 +112,9 @@ module.exports = function(
       };
 
       this.isActive = function() {
+        if (EditableFieldsState.forcedEditState) {
+          return false;
+        }
```
Comment (NobodysNightmare) [inline frontend/app/work_packages/directives/inplace_editor/inplace-editor-edit-pane-directive.js:117]: I have a bad suspicion here:
- I know that there is the requirement to not show the (✓, ✓ ✉, X) symbols when tabbing inside the create view
- I could imagine, that those buttons are displayed whenever the field is "active"
- Now there are two different conclusions from those two facts:
  1. Even in create mode a field may be active, but that does not neccessarily display the afforementioned buttons
  2. Apparently in create mode fields can't be active, luckily that means we will never show the buttons

The second conclusion is an evil one, because one could imagine other things depending on the "activeness" of a field that are valid in create and edit mode. However, I believe the author of the code above has chosen to follow that conclusion.

**tl; dr:** I believe the three lines above follow the following line of thought:
1. in create mode `forcedEditState` state is set (_I agree_)
2. `forcedEditState` means I am in create mode (**_I disagree**_)
3. in create mode a field can never be active, so that those fancy buttons are never shown (**_I disagree**_)

File: frontend/app/services/work-package-service.js (hunk 7 above)
Comment (floriank) [inline frontend/app/services/work-package-service.js:28]: this is a `jshint` comment with semantic function: If not used, @houndci would complain about `URI` being used without declaring it.

File: frontend/app/work_packages/controllers/work-package-new-controller.js (hunk 16 above)
Comment (floriank) [inline frontend/app/work_packages/controllers/work-package-new-controller.js]: `ViewModel`

File: frontend/app/templates/work_packages.list.new.html (hunk 2 above)
Comment (NobodysNightmare) [inline frontend/app/templates/work_packages.list.new.html]: It bothers me too...

File: frontend/app/work_packages/controllers/menus/tasks-dropdown-menu-controller.js (hunk 15 above)
Comment (floriank) [inline frontend/app/work_packages/controllers/menus/tasks-dropdown-menu-controller.js]: No, they are not.

File: frontend/app/templates/work_packages.list.new.html (hunk 2 above)
Comment (floriank) [inline frontend/app/templates/work_packages.list.new.html]: Well, it seems that this is due to the differences in styling of the different tabs. As the create tab does not have tabs, @0xF013 introduced a modification to deal with that fact.

It's a long term symptom of the _very different_ form views for create and edit.

File: frontend/app/api/hal-api-resource.js (hunk 4 above)
Comment (floriank) [inline frontend/app/api/hal-api-resource.js:64]: It is the same as above in `setup`. As the `HALAPIResource` object is of our own making this should be okay (for now).

I'll try to refactor a bit, as the `setupLink` method is imho a duplicate and `configure` should not be part of a public API to `HALAPIResource`

File: frontend/app/templates/work_packages.list.new.html (hunk 14 above)
Comment (floriank) [inline frontend/app/templates/work_packages.list.new.html]: I belive it should be possible to create a `details-toolbar` (Naming is hard) directive in the future which can hold the logic for all the views.

however, the current logic is packaged into `workPackageDetailsToolbar`, which I will not touch at the moment.

**Comment by ulferts** [comment]
This needs some rebasing to get it to the current 4.3, sorry @floriank.

File: frontend/app/work_packages/controllers/work-package-new-controller.js (hunk 17 above)
Comment (floriank) [inline frontend/app/work_packages/controllers/work-package-new-controller.js]: I'll refactor the `firstTimeFocused` part out, however, at the moment, there is little I can do for the unification of the activation code for the WP itself. The Overview (Show action) currently just listens to an event thrown (from someeeewheeeeere else). the create view actively intializes it's empty WP.

To rearange that I'd probably have to shove some abstraction in between, which, frankly, is beyond the scope of the PR.

**Comment by floriank** [comment]
@NobodysNightmare We talked about the closures stuff - overengineering on my part. 

the rest of your comments should be addressed, however, I will take larger refactoring into account for future iterations, as this PR is large enough as it is right now.

File: frontend/app/work_packages/controllers/work-package-new-controller.js (hunk 6 above)
Comment (floriank) [inline frontend/app/work_packages/controllers/work-package-new-controller.js]: `$stateParams` is part of the `angular` services around `ui-router`

**Comment by NobodysNightmare** [comment]
LGTM :+1: 

Lets wait for travis to finally agree...
